sql/analyzer/join_search.go

```diff
@@ -0,0 +1,495 @@
+	for i, accessOrder := range accessOrders {
+		cost := estimateTableOrderCost(tableNames, tablesByName, accessOrder, joinIndexes, lowestCost)
+		if cost < lowestCost {
+			lowestCost = cost
+			lowestCostIdx = i
+		}
+	}
+
+	cheapestOrder := make([]string, len(tableNames))
+	for i, j := range accessOrders[lowestCostIdx] {
+		cheapestOrder[i] = tableNames[j]
+	}
+
+	return cheapestOrder
+}
+
+// buildJoinTree builds a join plan for the tables in the access order given, using the join expressions given.
+func buildJoinTree(
+		tableOrder []string,
+		joinConds []*joinCond,
+) *joinSearchNode {
+
+	rootNodes := searchJoins(nil, &joinSearchParams{
+		tables:    tableOrder,
+		joinConds: joinConds,
+	})
+
+	for _, tree := range rootNodes {
+		// The search function here can return valid sub trees that don't have all the tables in the full join, so we need
+		// to check them for validity as an entire tree
+		if isValidJoinTree(tree) {
+			return tree
+		}
+	}
+
+	return nil
+}
+
+// Estimates the cost of the table ordering given. Lower numbers are better. Bails out and returns cost so far if cost
+// exceeds lowest found so far. We could do this better if we had table and key statistics.
+func estimateTableOrderCost(
+		tables []string,
+		tableNodes map[string]NameableNode,
+		accessOrder []int,
+		joinIndexes joinIndexesByTable,
+		lowestCost int,
+) int {
+	cost := 1
+	var availableSchemaForKeys sql.Schema
+	for i, idx := range accessOrder {
+		if cost >= lowestCost {
+			return cost
+		}
+
+		table := tables[idx]
+		availableSchemaForKeys = append(availableSchemaForKeys, tableNodes[table].Schema()...)
+		indexes := joinIndexes[table]
+
+		// If this table is part of a left or a right join, assert that tables are in the correct order. No table
+		// referenced in the join condition can precede this one in that case.
+		for _, idx := range indexes {
+			if (idx.joinType == plan.JoinTypeLeft && idx.joinPosition == plan.JoinTypeLeft) ||
+					(idx.joinType == plan.JoinTypeRight && idx.joinPosition == plan.JoinTypeRight) {
+				for j := 0; j < i; j++ {
+					otherTable := tables[accessOrder[j]]
+					if colsIncludeTable(idx.comparandCols, otherTable) {
+						return math.MaxInt32
+					}
+				}
+			}
+		}
+
+		if i == 0 || indexes.getUsableIndex(availableSchemaForKeys) == nil {
+			cost *= 1000
+		}
+	}
+
+	return cost
+}
+
+// colsIncludeTable returns whether the columns given contain the table given
+func colsIncludeTable(cols []*expression.GetField, table string) bool {
+	for _, col := range cols {
+		if strings.ToLower(col.Table()) == table {
+			return true
+		}
+	}
+	return false
+}
+
+// Generates all permutations of the slice given.
+func permutations(ints []int) [][]int{
+	var helper func([]int, int)
+	var res [][]int
+
+	helper = func(arr []int, n int){
+		if n == 1{
+			tmp := make([]int, len(arr))
+			copy(tmp, arr)
+			res = append(res, tmp)
+		} else {
+			for i := 0; i < n; i++{
+				helper(arr, n - 1)
+				if n % 2 == 1{
```

Comment:
This conditional is slightly mind bending...I might have expected something like...

```go
if n > len(arr) {
    res = append(res, append([]int(nil), arr...))
    return
}
helper(arr, n+1)
for i := n+1; i < len(arr); i++ {
    arr[n], arr[i] = arr[i], arr[n]
    helper(arr, n+1)
    arr[i], arr[n] = arr[n], arr[i]
}
```

for this whole function? Isn't it just: for each element in the remaining array, make it the current element and recurse with index one greater? Not trying to golf here and probably doesn't matter...but I think this is clearer:

```go
func permutations(a []int) (res [][]int) {
    var helper func(n int)
    helper = func(n int) {
        if n > len(a) {
            res = append(res, append([]int(nil), a...))
        } else {
            helper(n+1)
            for i := n+1; i < len(a); i++ {
                a[n], a[i] = a[i], a[n]
                helper(n+1)
                a[i], a[n] = a[n], a[i]
            }
        }
    }
    helper(0)
    return res
}

Reply:
You da man!